raiden_contracts/tests/test_channel_settle_unlock_state.py

```diff
@@ -0,0 +1,492 @@
+    (
+        expected_final_balance_A0,
+        expected_final_balance_B0,
+    ) = get_expected_after_settlement_unlock_amounts(vals_A0, vals_B0)
+
+    test_settlement_outcome(
+        (A, B),
+        (vals_A0, vals_B0, 'valid'),
+        expected_settlement0,
+        expected_settlement_onchain0,
+        expected_final_balance_A0,
+        expected_final_balance_B0,
+    )
+
+    if 'old_last' in channel_test_values:
+        for vals_A in channel_test_values['old_last']:
+            vals_B = vals_B0
+            test_settlement_outcome(
+                (A, B),
+                (vals_A, vals_B, 'old_last'),
+                expected_settlement0,
+                expected_settlement_onchain0,
+                expected_final_balance_A0,
+                expected_final_balance_B0,
+            )
+
+    if 'last_old' in channel_test_values:
+        for vals_B in channel_test_values['last_old']:
+            vals_A = vals_A0
+            test_settlement_outcome(
+                (A, B),
+                (vals_A, vals_B, 'last_old'),
+                expected_settlement0,
+                expected_settlement_onchain0,
+                expected_final_balance_A0,
+                expected_final_balance_B0,
+            )
+
+
+@pytest.mark.parametrize('channel_test_values', channel_settle_test_values)
+def test_channel_settle_old_old_balance_proof_values(
+        web3,
+        get_accounts,
+        assign_tokens,
+        channel_test_values,
+        create_channel_and_deposit,
+        test_settlement_outcome,
+):
+    (A, B, C, D) = get_accounts(4)
+    (vals_A0, vals_B0) = channel_test_values['valid_last']
+    assert are_balance_proofs_valid(vals_A0, vals_B0)
+    assert not is_balance_proof_old(vals_A0, vals_B0)
+
+    # Mint additional tokens for participants
+    assign_tokens(A, 400)
+    assign_tokens(B, 200)
+    # We make sure the contract has more tokens than A, B will deposit
+    create_channel_and_deposit(C, D, 40, 60)
+
+    expected_settlement0 = get_settlement_amounts(vals_A0, vals_B0)
+    expected_settlement_onchain0 = get_onchain_settlement_amounts(vals_A0, vals_B0)
+    (
+        expected_final_balance_A0,
+        expected_final_balance_B0,
+    ) = get_expected_after_settlement_unlock_amounts(vals_A0, vals_B0)
+
+    if 'old_last' in channel_test_values and 'last_old' in channel_test_values:
+        for vals_A in channel_test_values['old_last'][:5]:
+            for vals_B in channel_test_values['last_old'][:5]:
+                test_settlement_outcome(
+                    (A, B),
+                    (vals_A, vals_B, 'invalid'),
+                    expected_settlement0,
+                    expected_settlement_onchain0,
+                    expected_final_balance_A0,
+                    expected_final_balance_B0,
+                )
+
+
+@pytest.mark.parametrize('channel_test_values', channel_settle_test_values)
+def test_channel_settle_old_old_balance_proof_values2(
```

Comment:
As far as I can see this and the last two tests are identical except for the subsets of results they check. Can't we avoid the code duplication by turning the whole setup into a fixture, or turning the three tests into one parametrized one?

Reply:
About `test_channel_settle_valid_old_balance_proof_values`, `test_channel_settle_old_old_balance_proof_values` and `test_channel_settle_old_old_balance_proof_values2`:

- all the balance proof value pairs cannot be run together under the same test because travis fails with `no output for 10 min` and looking into this, I did not find a way to make travis wait except for `travis_wait`, which hides the logs until the job finishes (really not useful because you cannot pick up on failed tests early), so I just split the tested values because I thought the remaining code is not that much and did not want to add too many fixtures (add more magic to this)
- they both use `test_settlement_outcome` as a fixture for the bulk of the tests
- tried by making a fixture with the balance proof pairs, but the tests ended up running for > 2h - so we cannot do a test/test_value.

But I am open to suggestions here - I had a bit of a hard time to do the setup.

Reply:
If we don't want to rewrite much or add more fixtures, what comes to my mind is a single test function with an additional argument `tested_range` that is parametrized like this:
```
@pytest.mark.parametrize('tested_range', ('valid_old', 'old_old1', 'old_old2'))

# ... setup

if tested_range is 'valid_old':
    # ... assertions from the first test
elif tested_range is 'old_old1':
    # ... assertions from the second test
else:
    # ... assertions from the third test
```
A bit unusual but also well readable... Otoh the duplicated code is really not that much and can stay as it is instead.

Reply:
Is it ok for now to leave it as is and you can make a PR implementing the parametrized range after?

Reply:
Sure. I'll approve then.

Reply:
Opened https://github.com/raiden-network/raiden-contracts/issues/273 if you want to handle it.